cmd/config.go

```diff
@@ -0,0 +1,134 @@
+
+var nodeCmd = &cobra.Command{
+	Use:   "set-node",
+	Short: "Set the node to use for kubectl-az commands",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		node, err := cmd.Flags().GetString(utils.NodeKey)
+		if node == "" || err != nil {
+			return fmt.Errorf("node must be specified")
+		}
+		configs := make(map[string]string)
+		configs[utils.NodeKey] = node
+		return writeConfig(configs)
+	},
+}
+
+var resourceIDCmd = &cobra.Command{
+	Use:   "set-id",
+	Short: "Set the resource ID to use for kubectl-az commands",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		resourceID, err := cmd.Flags().GetString(utils.ResourceIDKey)
+		if resourceID == "" || err != nil {
+			return fmt.Errorf("id must be specified")
+		}
+
+		configs := make(map[string]string)
+		configs[utils.ResourceIDKey] = resourceID
+		return writeConfig(configs)
+	},
+}
+
+var vmmsCmd = &cobra.Command{
+	Use:   "set-vmss",
+	Short: "Set the VMSS node information to use for kubectl-az commands",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		subscription, err := cmd.Flags().GetString(utils.SubscriptionIDKey)
+		if subscription == "" || err != nil {
+			return fmt.Errorf("subscription must be specified")
+		}
+
+		nodeResourceGroup, err := cmd.Flags().GetString(utils.NodeResourceGroupKey)
+		if nodeResourceGroup == "" || err != nil {
+			return fmt.Errorf("node-resource-group must be specified")
+		}
+
+		vmss, err := cmd.Flags().GetString(utils.VMSSKey)
+		if vmss == "" || err != nil {
+			return fmt.Errorf("vmss must be specified")
+		}
+
+		instanceID, err := cmd.Flags().GetString(utils.InstanceIDKey)
+		if instanceID == "" || err != nil {
+			return fmt.Errorf("instance-id must be specified")
+		}
+
+		configs := make(map[string]string)
+		configs[utils.SubscriptionIDKey] = subscription
+		configs[utils.NodeResourceGroupKey] = nodeResourceGroup
+		configs[utils.VMSSKey] = vmss
+		configs[utils.InstanceIDKey] = instanceID
+
+		return writeConfig(configs)
+	},
+}
+
+var unsetCmd = &cobra.Command{
+	Use:   "unset",
+	Short: "Unset the configuration for kubectl-az",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return writeConfig(nil)
+	},
+}
+
+var showCmd = &cobra.Command{
+	Use:   "show",
+	Short: "Show the configuration for kubectl-az",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		fmt.Printf("%s: %s\n", utils.NodeKey, viper.GetString(utils.NodeKey))
+		fmt.Printf("%s: %s\n", utils.ResourceIDKey, viper.GetString(utils.ResourceIDKey))
+		fmt.Printf("%s: %s\n", utils.SubscriptionIDKey, viper.GetString(utils.SubscriptionIDKey))
+		fmt.Printf("%s: %s\n", utils.NodeResourceGroupKey, viper.GetString(utils.NodeResourceGroupKey))
+		fmt.Printf("%s: %s\n", utils.VMSSKey, viper.GetString(utils.VMSSKey))
+		fmt.Printf("%s: %s\n", utils.InstanceIDKey, viper.GetString(utils.InstanceIDKey))
+		return nil
+	},
+}
+
+func writeConfig(configs map[string]string) error {
+	// reset to ensure we always write a fresh config
+	viper.Reset()
```

Comment:
It means we are removing what was already written in the file, right? I'm not sure. I would find it annoying that setting the node removes the vmss-instance configuration. However, I don't know how to select the ones to use: node, vmss-instance or resource? Mmmm maybe another variable?

```bash
# It will first try to use the resource, if not present will try with the vmss-instance fields. Otherwise, try with node.
$ kubectl-az config set target-priority resource,vmss-instance,node

# To make it only try with the node
$ kubectl-az config set target-priority node
```

Of course, all the is valid only if user runs the command without flags:

```bash
# Use the set node
$ kubectl-az check-apiserver-connectivity --node myNode

# Use the env
$ NODE=myNode kubectl-az check-apiserver-connectivity

# Use the values in the config file with the order set in target-priority (default: vmss-instance, resource, node)
$ kubectl-az check-apiserver-connectivity

# Use the value of node in the config file
$ kubectl-az config set target-priority node
$ kubectl-az check-apiserver-connectivity
```

In case we add other config values not related with the target resource (or maybe we could already do it?), we could group all this in one group:

```bash
$ kubect-az config show
target:
    priority: vmss-instance,resource,node
    vmss-instance:
        subscription: mySub
        node-resource-group: myNRG
        vmss: myVMSS
        id: myInstanceID
    resource:
        id: /subscriptions/mySub/resourceGroups/myNRG/providers/Microsoft.Compute/virtualMachineScaleSets/myVMSS/virtualMachines/myInstanceID
    node:
        name: Node1
```

WDYT? Is it too much? Am I over-engineering it?

Reply:
I think the concern is valid but I feel like `target-priority` might be over-engineering? like for the initial version we can print some warning that multiple config found and using `node` as default. User can still unset after warning and if flags/env are set they take priority. What do you say?

Reply:
Yep, maybe a warning is fine for now. However, the default shouldn't be `node` but `vmss-instance` or `resource`, as they will work also in case the Kubernetes control plane is not working properly. With `node` we need to make a call to the API to translate the node name to the vmss instance information. So, the priority should be `vmss-instance` or `resource` > node.